File: interfaces/src/main/java/org/codroid/interfaces/preference/PreferencesProperty.java
package org.codroid.interfaces.preference;

import com.tencent.mmkv.MMKV;

import org.codroid.interfaces.env.AddonEnv;
import org.codroid.interfaces.env.Property;
import org.codroid.interfaces.utils.PathUtils;
import org.codroid.interfaces.utils.TomlKt;

import java.nio.file.Path;
import java.util.function.BiConsumer;

import cc.ekblad.toml.TomlMapper;

public class PreferencesProperty extends Property {

    private final MMKV mmkv;
    private Preferences preferences;

    public PreferencesProperty(Path path) {
        super(path);
        this.mmkv = MMKV.mmkvWithID(path.getFileName().toString() + "-kv", path.toAbsolutePath().toString());
        initDefaultValues();
    }

    public PreferencesProperty(AddonEnv addonEnv, String relativePathStr) {
        super(addonEnv, relativePathStr);
        this.mmkv = MMKV.mmkvWithID(addonEnv.getIdentify() + "-kv", addonEnv.getAddonRootDir().getAbsolutePath());
        initDefaultValues();
    }

    private void initDefaultValues() {
        this.preferences = getAs(Preferences.class);
        this.preferences.getSettings().forEach((k, v) -> {
            if (v instanceof InputSetting) {
                if (((InputSetting) v).getValueType().equals("STRING")) {
                    putString(k, ((InputSetting) v).getValueType());
                } else {
                    putInt(k, Integer.parseInt(((InputSetting) v).getDefaultValue().toString()));
                }
            } else if (v instanceof TextareaSetting) {
                putString(k, ((TextareaSetting) v).getDefaultValue());
            } else if (v instanceof SwitchSetting) {
                putBoolean(k, ((SwitchSetting) v).getDefaultValue());
            } else if (v instanceof SelectSetting) {
                putInt(k, ((SelectSetting) v).getDefaultValue());
            }
        });
    }

    public String getString(String key) {
        return this.mmkv.decodeString(key);
    }

    public int getInt(String key) {
        return this.mmkv.decodeInt(key);
    }

    public boolean getBoolean(String key) {
        return this.mmkv.decodeBool(key);
    }

    public void putString(String key, String value) {
        this.mmkv.encode(key, value);
    }

    public void putInt(String key, int value) {
        this.mmkv.encode(key, value);
    }

    public void putBoolean(String key, boolean value) {
        this.mmkv.encode(key, value);
    }

    @Override
    public TomlMapper getTomlMapper() {
        return PreferencesKt.getPreferencesMapper();
    }
}

File: interfaces/src/main/java/org/codroid/interfaces/preference/PreferencesProperty.kt
package org.codroid.interfaces.preference

import cc.ekblad.toml.TomlMapper
import com.tencent.mmkv.MMKV
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.codroid.interfaces.env.AddonEnv
import org.codroid.interfaces.env.Property
import java.io.InputStream
import kotlin.reflect.typeOf

class PreferencesProperty : Property<Preferences> {

    private val mmkv: MMKV

    constructor(addonEnv: AddonEnv, relativePath: String) : super(
        addonEnv,
        relativePath,
        typeOf<Preferences>()
    ) {
        mmkv = MMKV.mmkvWithID(addonEnv.identify + "-kv", addonEnv.addonRootDir.absolutePath)
        init()
    }

    constructor(id: String, mmkvPath: String, inputStream: InputStream) : super(
        inputStream,
        typeOf<Preferences>()
    ) {
        mmkv = MMKV.mmkvWithID(id, mmkvPath)
        init()
    }

    fun init() {
        entity?.settings?.forEach { (k, v) ->
            when (v) {
                is InputSetting -> {
                    if (v.valueType == "STRING") {
                        putString(k, v.valueType)
                    } else {
                        putInt(k, v.defaultValue.toInt())
                    }
                }
                is TextareaSetting -> putString(k, v.defaultValue)
                is SwitchSetting -> putBoolean(k, v.defaultValue)
                is SelectSetting -> putInt(k, v.defaultValue)
            }
        }
    }

    fun putString(key: String, value: String) {
        mmkv.encode(key, value)
    }

    fun putBoolean(key: String, value: Boolean) {
        mmkv.encode(key, value)
    }

    fun putInt(key: String, value: Int) {
        mmkv.encode(key, value)
    }

    fun getString(key: String) = mmkv.decodeString(key)

    fun getBoolean(key: String) = mmkv.decodeBool(key)

    fun getInt(key: String) = mmkv.decodeInt(key)

    override fun getTomlMapper(): TomlMapper {
        return preferencesMapper
    }

    @Serializable
    data class A(val string: String)

    fun toJson(): String {
        if (entity != null) {
            return Json.encodeToString(entity)
        }
        return ""
    }
}